fleece/parameters.py

```diff
@@ -0,0 +1,130 @@
+                encrypted = True
+                data = data[9:]
+            param_list.append((key_prefix, data, encrypted))
+        elif isinstance(data, list):
+            i = 0
+            for item in data:
+                flatten_item(item, '{}.{}'.format(key_prefix, i))
+                i += 1
+        elif isinstance(data, dict):
+            for k, v in data.items():
+                flatten_item(v, key_prefix + '/' + k)
+
+    flatten_item(config)
+    return param_list
+
+
+def _unflatten_config(param_list):
+    """Reconstructs a config dictionary from a list of key/value tuples."""
+    config = {}
+
+    def parse_key(key):
+        k = key
+        i = -1
+        if '.' in key:
+            k, i = key.split('.')
+            i = int(i)
+        return k, i
+
+    for param in param_list:
+        keys = param[0].split('/')
+
+        c = config
+        for k in keys[1:-1]:
+            key, index = parse_key(k)
+            if key not in c:
+                if index == -1:
+                    c[key] = {}
+                else:
+                    c[key] = []
+            c = c[key]
+            if index != -1:
+                while (len(c) < index + 1):
+                    c.append(None)
+                if c[index] is None:
+                    c[index] = {}
+                c = c[index]
+        key, index = parse_key(keys[-1])
+        if index == -1:
+            c[key] = param[1]
+        else:
+            if key not in c:
+                c[key] = []
+            c = c[key]
+            while (len(c) < index + 1):
+                c.append(None)
+            c[index] = param[1]
+
+    return config
+
+
+def read_parameters(root_path, **boto_args):
+    """Read parameters from EC2 Parameter Store previously written with the
+    write_parameters() function and return them as a configuration dictionary.
+    """
+    ssm = boto3.client('ssm', **boto_args)
+    r = ssm.describe_parameters(ParameterFilters=[{
+        'Key': 'Path',
+        'Option': 'Recursive',
+        'Values': [root_path]
+    }])
+    p = ssm.get_parameters(
+        Names=[param['Name'] for param in r['Parameters']],
+        WithDecryption=True)
+    param_list = [(param['Name'], param['Value']) for param in p['Parameters']]
+    return _unflatten_config(param_list)[root_path[1:]]
+
+
+def erase_parameters(root_path, **boto_args):
+    """Erase all parameters under the given path in the EC2 Parameter Store."""
+    ssm = boto3.client('ssm', **boto_args)
+    r = ssm.describe_parameters(ParameterFilters=[{
+        'Key': 'Path',
+        'Option': 'Recursive',
+        'Values': [root_path]
+    }])
+    for param in r['Parameters']:
+        ssm.delete_parameter(Name=param['Name'])
+
+
+def write_parameters(config, kms_key, root_path, **boto_args):
+    """Write a config file dictionary to the EC2 Parameter Store."""
+    param_list = _flatten_config(config)
+    ssm = boto3.client('ssm', **boto_args)
+
+    # first remove old parameters
```

Comment:
Why is it necessary to delete old parameters before setting the new ones? I'd have to double-check, but I _think_ `put_parameter` can be used for updates as well.

Reply:
What if a parameter was removed from the config file? This ensures that there are no left overs from a previous session.

Reply:
I also have some concerns around flushing everything, then uploading. This could negatively effect in-progress operations. Would it be possible to just remove missing ones? Ideally, the path should be based on the Lambda version anyway, which should prevent most of the problem, but I suppose I could see an issue where the config is updated and the Lambda is not.

Reply:
That's an interesting point. Adding a version to the config path makes a lot of sense in our use case. The lambda sweeper could clean up the configuration when it purges the corresponding lambda.

As a side note, I'm not really that excited about parameter store, tbh. Seems appropriate if you want to share configuration across different services, but in our case we will always have a single consumer, so I honestly don't see the benefit versus storing the encrypted config inside the lambda as a yaml file. And this would prevent the versioning issues.

Reply:
From a security standpoint, there are concerns with storing secrets using that method as AWS doesn't recommend storing sensitive data this way (despite the Lambda code being encrypted at rest). SSM Parameter store is a bit of a hack around storing encrypted environment variables and then calling KMS to decrypt each of them since it allows a batch get with decryption.

Reply:
I mean storing the encrypted config with the lambda. This would be a big improvement over our current method with yoke, which stores the plain text secrets. The `load_config()` function would do the decrypting at run-time.

Reply:
Ah, I missed the 'encrypted' part of `storing the encrypted config inside the lambda`. How would you propose decrypting the config? The only ways I can think of are:

1) Hard limit of 4K and use one KMS call to decrypt the file.
2) Decrypt all values individually - this requires a lot of KMS API calls, which can be sped up by running async, but still adds overhead to a Lambda.
3) Storing an encrypted data key within the encrypted config and using one KMS call to decrypt the data key, then use `cryptography` in the Lambda to decrypt the config. This bypasses the 4K limit, but I don't know what the overhead of requiring `cryptography` and decrypting locally in the Lambda would be. Surely lower memory Lambda's (and thus with lower CPU) would be more negatively affected than others. I don't have any data around this though.

Reply:
How about the entire json or yaml payload is broken up in chunks of 4K that are encrypted/decrypted with the KMS key? For most services you'll get everything with a single decrypt call.

Reply:
Based on all the work to do SSM, that sounds appealing. If we went that route, then I think the `yoke` style config might be best too. I would be good to hear from @larsbutler @szilveszter and @Gifflen on this though.

Reply:
So there are tradeoffs between doing the per item encryption and per file encryption. The per item helps significantly for cleaning up PRs and making it a more pleasant experience to work on as a dev and actually tracking changes over time. The per file method would have less impact on performance when it comes to loading the config in the lambda.

My thought is it might make sense to have both. If we use the per key in encryption to keep the file sane in source control but in CI when we go to bundle the zip inject a single encrypted blob for the lamba to use. I don't think this should add too much overhead since we would need the encrypt/decrypt bits regardless.

One of the draw backs is not using the exact same exact config in source control as the one being used by the lambda but that is a relatively minor concern that I don't think outweighs the benefits of this 2 prong approach.

Thoughts?

Reply:
> One of the draw backs is not using the exact same exact config in source control as the one being used by the lambda

This is true anyway. The config file will have values for all stages, while the version of the config that would end up in the lambda would be a "rendered" version that only has the values for the appropriate stage.